tiledb/api/c_api/vfs/vfs_api.cc

```diff
@@ -325,7 +325,17 @@ capi_return_t tiledb_vfs_ls_recursive(
         "Invalid TileDB object: Callback function is null.");
   }
   ensure_output_pointer_is_valid(data);
-  vfs->ls_recursive(tiledb::sm::URI(path), callback, data);
+
+  tiledb::sm::LsObjects lsresults;
+  throw_if_not_ok(vfs->ls_recursive(tiledb::sm::URI(path), &lsresults));
```

Comment:
For the reviewers:

This change was one of the ways to get the empty subdirectories to appear in the *external* API output (they appear just fine in the internal API output).

Previously the callback supplied here found its way down the call stack and was being used as the `file_filter` argument to the internal API.  We wouldn't see empty directories (or any directories) for that, of course, since the the `file_filter` is not invoked on them.  The return value of the internal `ls_recursive` was dropped on the floor and the regular file results were duplicated in the callback state.

I assume the intent of putting the callback in this way was so that the C API callback could be interleaved with the `ls` expansion?  IMO if we want to continue enabling that the best way would be to propagate the separate file/directory filter callbacks all the way up to the C API.

Reply:
Yeah, we haven't added support for pruning directories with the directory_filter yet. Internally as I'm sure you've noticed this argument defaults to `tiledb::sm::accept_all_dirs` so all directories are always accepted, but I see what you are saying. The results are included in the value returned from ls_recursive but not the data managed by the callback. This isn't the case for S3 since there is no concept of directories on S3, we apply the callback to all objects. 

I think for now the best way forward is to call the one user supplied callback for directories and files, just pass in a `0` for the object size here. Adding support for the directory_filter could be done separately IMO.

https://github.com/TileDB-Inc/TileDB/blob/22b2e3b719a9cccc2932981c8553fb5bd5bc04a6/tiledb/sm/filesystem/posix.h#L389

If we do this, `ls_recursive("root/")` on a test tree with `root/subdir/` and `root/subdir/file.txt`, both of these will be included in the results. This is what I would expect on posix but worth double-checking with @KiterLuc or @eric-hughes-tiledb on this. If we want to filter out the non-empty directories we could do that with `std::filesystem::is_empty`.

Reply:
I changed the internal API to have a single `PathPredicate P` and made the common case for that be `PathFilter = std::function<PathFilterResult(const directory_entry_uri_view &)>` which unifies a bunch of scenarios and adds a lot of flexibility as to how the iteration proceeds.  This is a pretty sizable change but it should be covered pretty well by the unit tests.

This also means that decisions about whether to filter empty directories, etc can be pushed closer to the external API rather than embedding them in the directory iteration function.

Reply:
Discussion this morning: we will revert several changes here in favor of landing a minimal change set in the imminent release.  The changes here will be re-homed to another branch where we can make more substantive internal and external API changes in the future.

For this comment in particular we're going to resolve it by calling `file_filter` as well on directories.